src/autogenerated/java/net/minestom/server/particle/ParticleType.java

```diff
@@ -0,0 +1,292 @@
+package net.minestom.server.particle;
+
+import net.kyori.adventure.key.Key;
+import net.kyori.adventure.key.Keyed;
+import net.minestom.server.particle.data.BlockParticle;
+import net.minestom.server.particle.data.DustParticle;
+import net.minestom.server.particle.data.ItemParticle;
+import net.minestom.server.particle.data.Particle;
+import net.minestom.server.registry.Registry;
+import net.minestom.server.utils.NamespaceID;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.function.BiFunction;
+
+/**
+ * AUTOGENERATED by ParticleTypeGenerator
+ */
+public class ParticleType<T extends Particle> implements Keyed {
```

Comment:
I am unsure about the decision to use a generic here, in which situation would it help?

Reply:
The generic is there to make sure you can't create, for example, a block particle effect with an item particle. You can only create particles without data if the type is without data, and you can only create data if the particle type needs that data.

Reply:
Perhaps that it would make more sense to store the particle data inside the type, so the particle effect contains the particle effect and additional information such as position/count. Would also allow better parsing for particle data and the ability to expand it with mods.

Reply:
When you have the particle data inside the type, how would you change the data? There can be different Particles, with different data but the same type, so how would that work?

Reply:
The Particle class is what really represents a particle, ParticleType is just the type

Reply:
Could add some methods to create a new `Particle` object with the required data